core/src/main/java/org/elasticsearch/index/store/Store.java

```diff
@@ -716,6 +719,38 @@ public String toString() {
             return "store(" + in.toString() + ")";
         }
 
+        @Override
+        public void copyFrom(Directory from, String srcFile, String destFile, IOContext context) throws IOException {
+            Directory fromUnwrapped = FilterDirectory.unwrap(from);
+            Directory toUnwrapped = FilterDirectory.unwrap(this);
+            // try to unwrap to FSDirectory - we might be able to just create hard-links of these files and safe copying
+            // the entire file.
+            if (fromUnwrapped instanceof FSDirectory
+                && toUnwrapped instanceof FSDirectory) {
+                final Path fromPath = ((FSDirectory)fromUnwrapped).getDirectory();
+                final Path toPath = ((FSDirectory)toUnwrapped).getDirectory();
+                try {
+                    Files.createLink(toPath.resolve(destFile), fromPath.resolve(srcFile));
+                } catch (FileNotFoundException | NoSuchFileException | FileAlreadyExistsException ex) {
+                    throw ex; // in these cases we bubble up since it's a true error condition.
+                } catch (IOException
+                    | UnsupportedOperationException // if the FS doesn't support hard-links
+                    | SecurityException ex // we might miss a java.nio.file.LinkPermission" "hard"
+                    ) {
+                    // hard-links are not supported or the files are on different filesystems
+                    // we could go deeper and check if their filesstores are the same and opt
+                    // out earlier but for now we just fall back to normal file-copy
+                    try {
+                        super.copyFrom(from, srcFile, destFile, context);
+                    } catch (Exception e) {
+                        e.addSuppressed(ex);
+                        throw ex;
+                    }
+                }
+            } else {
+                super.copyFrom(from, srcFile, destFile, context);
+            }
+        }
     }
```

Comment:
I think we had the discussion already on the Lucene mailing lists. I have no strong feelings, but hard links are often something bad to system administrators. So there should be at least some way to opt out.

I could think about having that on the Lucene side, but it would need the same strange workaround. unwrapDirectory is a no-go inside Lucene, sorry, because it may break very easy for customized or filtered directories.